File: mobile-app/android/app/src/main/java/com/taleemaateislam/app/NotificationTopicsPlugin.java
package com.taleemaateislam.app;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.google.firebase.messaging.FirebaseMessaging;

@CapacitorPlugin(name = "NotificationTopics")
public class NotificationTopicsPlugin extends Plugin {

    @PluginMethod
    public void subscribe(PluginCall call) {
        String topic = call.getString("topic");
        if (topic == null || topic.isEmpty()) {
            call.reject("topic is required");
            return;
        }
        FirebaseMessaging.getInstance().subscribeToTopic(topic).addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                call.resolve();
            } else {
                call.reject("Failed to subscribe to " + topic, task.getException());
            }
        });
    }

    @PluginMethod
    public void unsubscribe(PluginCall call) {
        String topic = call.getString("topic");
        if (topic == null || topic.isEmpty()) {
            call.reject("topic is required");
            return;
        }
        FirebaseMessaging.getInstance().unsubscribeFromTopic(topic).addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                call.resolve();
            } else {
                call.reject("Failed to unsubscribe from " + topic, task.getException());
            }
        });
    }
}

File: mobile-app/ios/App/App/NotificationTopicsPlugin.swift
import Capacitor
import FirebaseMessaging

// Mirrors android/app/src/main/java/com/taleemaateislam/app/NotificationTopicsPlugin.java -
// same JS-facing name/methods so components.js needs no platform branching.
@objc(NotificationTopicsPlugin)
public class NotificationTopicsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NotificationTopicsPlugin"
    public let jsName = "NotificationTopics"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "subscribe", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "unsubscribe", returnType: CAPPluginReturnPromise)
    ]

    @objc func subscribe(_ call: CAPPluginCall) {
        guard let topic = call.getString("topic"), !topic.isEmpty else {
            call.reject("topic is required")
            return
        }
        Messaging.messaging().subscribe(toTopic: topic) { error in
            if let error = error {
                call.reject("Failed to subscribe to \(topic)", nil, error)
            } else {
                call.resolve()
            }
        }
    }

    @objc func unsubscribe(_ call: CAPPluginCall) {
        guard let topic = call.getString("topic"), !topic.isEmpty else {
            call.reject("topic is required")
            return
        }
        Messaging.messaging().unsubscribe(fromTopic: topic) { error in
            if let error = error {
                call.reject("Failed to unsubscribe from \(topic)", nil, error)
            } else {
                call.resolve()
            }
        }
    }
}
